### PR TITLE
Implementation of PseudoSensor

### PR DESCRIPTION
**Pull request by alxschwrz**
Added implementation of PseudoSensor. PseudoSensor observes obstacles and publishes information to the observation_space. Still some features missing. goal observation not implemented yet. PseudoDistSensor not implemented yet. Fixed setSpaces issue in pointRobotEnv when calling resetLimits() (ref: #32)

### CODE REVIEW

File: examples/point_robot.py [hunk 1]
```diff
@@ -1,6 +1,9 @@
 import gym
 import planarenvs.pointRobot
 import numpy as np
+from sensors.sensors import PseudoSensor
+from sensors.sensors import PseudoDistSensor
+from sensors.sensors import GoalSensor
```
Comment (maxspahn) [inline examples/point_robot.py:6]: Could you make them optional imports?
Just like with the obstacles. Then the imports are only performed if sensors-flag is set to 1.

File: examples/point_robot.py [hunk 2]
```diff
@@ -9,15 +12,23 @@
...
+    goalPosObserver = GoalSensor(nbGoals=1, mode='position')
+    env.addSensor(goalPosObserver)
+
     defaultAction = lambda t: np.array([np.cos(1.0 * t), np.sin(1.0 * t)])
     initPos = np.array([0.0, -1.0])
     initVel = np.array([-1.0, 0.0])
-    n_episodes = 1
-    n_steps = 1000
+    n_episodes = 2
+    n_steps = 10000
```
Comment (maxspahn) [inline examples/point_robot.py:27]: Undo this change. Not required for this specific feature.

File: planarenvs/pointRobot/envs/pointRobotEnv.py [hunk 3]
```diff
@@ -36,7 +37,14 @@ def resetLimits(self, **kwargs):
...
                     import logging
                     logging.warning("Ignored reset of limit because the size of the limit is incorrect.")
-        self.setSpaces()
+
+        self._limUpPos = self._limits['pos']['high']
+        self._limUpVel = self._limits['vel']['high']
+        self._limUpAcc = self._limits['acc']['high']
+        self._limUpFor = self._limits['for']['high']
+        self.observation_space['x'] = spaces.Box(low=-self._limUpPos, high=self._limUpPos, dtype=np.float64)
+        self.observation_space['xdot'] = spaces.Box(low=-self._limUpVel, high=self._limUpVel, dtype=np.float64)
```
Comment (maxspahn) [inline planarenvs/pointRobot/envs/pointRobotEnv.py:46]: For future issue: the `self._limUpPos` should be entirely replaced by` self._limits[...]['high'/'low']`.
Not part of this PR so fine.

**Review by maxspahn — CHANGES_REQUESTED** [review]
Nice work. There are some minor changes and one major change concerns hierarchy for the sensors. See comment in the sensors.py file. Just comment the comments if you have question on them.

**Comment by alxschwrz** [comment]
Thanks a lot for the input. I will work on your change requests and let you know as soon as I have implemented the requested changes.

File: planarenvs/planarCommon/planarEnv.py [hunk 4]
```diff
@@ -71,11 +80,17 @@ def reset(self, pos=None, vel=None):
...
         return self._get_ob()
 
     def step(self, a):
         self.action = a
         self.integrate()
+        for sensor in self._sensors:
+            if sensor.name() in ["GoalPosition", "GoalDistance"]:
+                self.state[sensor.name()] = sensor.sense(self.state, self._goals, self.t())
+            else:
+                self.state[sensor.name()] = sensor.sense(self.state, self._obsts, self.t())
```
[the remark this answers is not in the document]
Comment (alxschwrz) [reply planarenvs/planarCommon/planarEnv.py:93]: I will pass both obstacles and goals to the `sense` function.

**Comment by alxschwrz** [comment]
Hi @maxspahn I included the changes and merged the new master branch into this feature branch. Could you check the updated PR?

File: planarenvs/planarCommon/planarEnv.py [hunk 5]
```diff
@@ -76,6 +85,8 @@ def reset(self, pos=None, vel=None):
     def step(self, a):
         self.action = a
         self.integrate()
+        for sensor in self._sensors:
+            self.state[sensor.name()] = sensor.sense(self.state, self._goals, self._obsts, self.t())
```
Comment (maxspahn) [inline planarenvs/planarCommon/planarEnv.py:89]: The function call is great and exactly as I hoped it would be. Passing `self._goals` and `self._obsts` seems like the better solution here.
I would prefer to not add it to `self.state` as it might lead to trouble in the integration scheme. Could make a second attribute to the `planarEnv` class that stored sensed information? You then only concatenate them in the function `self._get_ob()`.

File: planarenvs/pointRobot/envs/pointRobotEnv.py [hunk 6]
```diff
@@ -36,7 +37,14 @@ def resetLimits(self, **kwargs):
...
                     logging.warning("Ignored reset of limit because the size of the limit is incorrect.")
-        self.setSpaces()
+
+        self._limUpPos = self._limits['pos']['high']
+        self._limUpVel = self._limits['vel']['high']
+        self._limUpAcc = self._limits['acc']['high']
+        self._limUpFor = self._limits['for']['high']
+        self.observation_space.spaces['x'] = spaces.Box(low=-self._limUpPos, high=self._limUpPos, dtype=np.float64)
+        self.observation_space.spaces['xdot'] = spaces.Box(low=-self._limUpVel, high=self._limUpVel, dtype=np.float64)
+
```
Comment (maxspahn) [inline planarenvs/pointRobot/envs/pointRobotEnv.py:47]: Refer to #32 
I will fix this at a later stage.

File: sensors/GoalSensor.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
+from sensors.SensorCommon import Sensor
+from sensors.SensorCommon import Dist2Circ
+import numpy as np
+import operator
+
+
+class GoalSensor(Sensor):
+
+    def __init__(self, nbGoals=0, SensorRange=10, mode="position"):
```
Comment (maxspahn) [inline sensors/GoalSensor.py:9]: What is the purpose fo this SensorRange variable? I understand it as being the default value when no observation is made (no obstacle for example). Then the naming is confusing.
Could you also be consistent with variables and functions being not-capatilized?

File: sensors/GoalSensor.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+                    break
+                currGoalPos = goal.position(t=t)
+                currGoalDist = Dist2Circ(s['x'][0], s['x'][1], currGoalPos[0], currGoalPos[1], goal.epsilon())
+                self._Observation[idx][0] = currGoalDist[0]
+                self._Observation[idx][1] = currGoalDist[1]
+
+        return self._Observation.clip(-self._limSensor, self._limSensor)
+
+
+class OldGoalSensor(Sensor):
```
Comment (maxspahn) [inline sensors/GoalSensor.py:48]: I guess this is outdated now?

File: sensors/SensorCommon.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        pass
+
+    def getOSpaceSize(self):
+        return self._nbObservations, 2
+
+    def getObservationSpace(self):
+        return spaces.Box(low=-self._limSensor, high=self._limSensor, shape=self.getOSpaceSize(), dtype=np.float64)
+
+    @abstractmethod
+    def _reset(self):
```
Comment (maxspahn) [inline sensors/SensorCommon.py:35]: More out of interest: Are you following some specific guide (PEP rules for example) or why are some methods started with `_` (underscore)?

File: sensors/SensorCommon.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+def Dist2Circ(x1, y1, x2,  y2, r):
+    dx = x2 - x1
+    dy = y2 - y1
+    Dist2Center = (dx ** 2 + dy ** 2) ** (1 / 2)
+    if Dist2Center < r:
+        return np.zeros(2)
+    dx = dx * (Dist2Center - r) / Dist2Center
+    dy = dy * (Dist2Center - r) / Dist2Center
+    return dx, dy
```
Comment (maxspahn) [inline sensors/SensorCommon.py:14]: Make variables and functions lowercase to not confuse them with classes.
Would it be possible to pass [x1, y1] as array directly and not have the explicit splitting between the two variables?
That would make it easier later to extend to 3D if we want to and it makes the code shorter and cleaner.

File: sensors/ObstacleSensor.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+        self._Observation[:] = self._limSensor
+
+    def sense(self, s, goals, obstacles, t=0):
+        self._reset()
+        if self._mode == "position":
+            for idx, obst in enumerate(obstacles):
+                if idx >= self._nbObservations:
+                    break
+                self._Observation[idx][0] = obst.position(t=t)[0]
+                self._Observation[idx][1] = obst.position(t=t)[1]
```
Comment (maxspahn) [inline sensors/ObstacleSensor.py:35]: Do you need to explicitly split between x and y coordinate?

File: sensors/ObstacleSensor.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+        elif self._mode == "distance":
+            self._reset()
+            currPos = s['x']  # only true for point_robot
+            obstaclesInfo = []
+            for idOb, obstacle in enumerate(obstacles):
+                currObstPos = obstacle.position(t=t)
+                currObstDist = Dist2Circ(currPos[0], currPos[1], currObstPos[0], currObstPos[1], obstacle.radius())
```
Comment (maxspahn) [inline sensors/ObstacleSensor.py:42]: see above

File: sensors/ObstacleSensor.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+        ## todo: implement obstacle and goal flag.
+        self._Observation = np.ones([self._nbObservations, 2]) * self._limSensor
+        self._mode = mode
+        self._setSensorName()
+
+    def _setSensorName(self):
+        if self._mode == "position":
+            self._name = "ObstaclePseudoSensorPosition"
+        elif self._mode == "distance":
+            self._name = "ObstaclePseudoSensorDistance"
```
Comment (maxspahn) [inline sensors/ObstacleSensor.py:20]: This name is a bit too long for a dict-key. Also for consistency with respect to the GoalSensor you can call it `ObstaclesPositions`.

File: sensors/GoalSensor.py [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+        self._Observation[:] = 0
+
+    def sense(self, s, goals, obstacles, t=0):
+        self._reset()
+        if self._mode == "position":
+            for idx, goal in enumerate(goals):
+                if idx >= self._nbObservations:
+                    break
+                self._Observation[idx][0] = goal.position(t=t)[0]
+                self._Observation[idx][1] = goal.position(t=t)[1]
```
Comment (maxspahn) [inline sensors/GoalSensor.py:34]: See comment about explicit splitting between x and y coordinate in ObstacleSensor.

File: sensors/GoalSensor.py [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+
+        elif self._mode == "distance":
+            for idx, goal in enumerate(goals):
+                if idx >= self._nbObservations:
+                    break
+                currGoalPos = goal.position(t=t)
+                currGoalDist = Dist2Circ(s['x'][0], s['x'][1], currGoalPos[0], currGoalPos[1], goal.epsilon())
```
Comment (maxspahn) [inline sensors/GoalSensor.py:41]: The variable `s` refers to the current state I assume. Maybe you can think of a more meaningful name here.
Then, you refer directly to the state as the forward kinematics of the point robot is the state essentially (configuration space equals workspace). Therefore, this does not work for other robots at the moment. 
We need to either pass the forward kinematics to the sensor on initialize forward kinematics in the sensor constructor. 
The second option seems better, as the goal object already contains information about which link the goal is defined for. We can discuss this later.

**Review by maxspahn — CHANGES_REQUESTED** [review]
Great work! I tested it (only with the point robot so far) and it works like a charm!
I made some comments mostly about variable naming and capitalized letters.

In the current stage, this implementation only supports point robots, as the forward kinematics is assumed to be the states of the robot. When experimenting with other kinematic chains, this implementation will genarilize to all robots.

File: sensors/GoalSensor.py (hunk 8 above)
Comment (alxschwrz) [reply sensors/GoalSensor.py:48]: yes it is! Thanks!

File: sensors/SensorCommon.py (hunk 9 above)
Comment (alxschwrz) [reply sensors/SensorCommon.py:35]: My idea behind that was to make methods, which are meant for internal use only (kind of like private methods), starting with `_` (underscore). Is this the same way you are using the `_` in the codebase? Would be nice to stay consistent with your style. 

Also, do you think the _reset() methods for the sensors should be only for internal use or also external?

File: sensors/GoalSensor.py (hunk 7 above)
Comment (alxschwrz) [reply sensors/GoalSensor.py:9]: Exactly, it is meant to be the default value when no observation is made. Since "zero-padding" might be not the best idea for sensor information I figured it might be best to be able to set it as intended. I will find a more suitable name for it!
Also, I will make naming in general more consistent!

File: sensors/SensorCommon.py (hunk 10 above)
Comment (alxschwrz) [reply sensors/SensorCommon.py:14]: will do. good point, I will change the function to take arrays.

File: sensors/ObstacleSensor.py (hunk 11 above)
Comment (alxschwrz) [reply sensors/ObstacleSensor.py:35]: no not at all. I can pass them directly.

File: sensors/GoalSensor.py (hunk 15 above)
Comment (alxschwrz) [reply sensors/GoalSensor.py:41]: Regarding the naming: Definetely!
kinematic chain: Exactly, right now it only works for the point robot. Would love to discuss how we can make this general for all robots!

File: sensors/GoalSensor.py (hunk 7 above)
Comment (alxschwrz) [reply sensors/GoalSensor.py:9]: Sorry, I mixed up the answer here. SensorRange is not the default value for when no observation is made, but the limit for the observation space. 
For the ObstacleSensor, this aligns with the default value when no obstacle is present. For the GoalSensor, the default value is set to 0. The default value is set in the _reset function.

File: sensors/SensorCommon.py (hunk 9 above)
Comment (maxspahn) [reply sensors/SensorCommon.py:33]: The issue is that I have no consistent style at the moment. I am style looking for one. :D
We should start looking into PEP8 guidelines as it might help.

Concerning this specific point: You can keep it as `_reset()` and then we could clean up once we settled on one style guide.

File: sensors/SensorCommon.py [hunk 16]
```diff
@@ -0,0 +1,38 @@
+from abc import ABC, abstractmethod
+import numpy as np
+from gym import spaces
+
+
+def dist2circ(pointPos, circlePos, circleRadius):
+    diff = pointPos - circlePos
+    dist2center = (diff[0] ** 2 + diff[1] ** 2) ** (1 / 2)
```
Comment (maxspahn) [inline sensors/SensorCommon.py:8]: Simply use `np.linalg.norm(diff)`
It will avoid struggle when we want to move to higher orders, although it is unlikely to happen.

File: sensors/ObstacleSensor.py [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+from sensors.SensorCommon import dist2circ
+import numpy as np
+import operator
+
+
+class ObstacleSensor(Sensor):
+
+    def __init__(self, obstacles=False, nbObstacles=0, limSensor=10, mode="position"):
+        super().__init__(nbObservations=nbObstacles, limSensor=limSensor)
+        ## todo: implement obstacle and goal flag.
```
Comment (maxspahn) [inline sensors/ObstacleSensor.py:11]: I guess this is obsolete by now?

File: planarenvs/planarCommon/planarEnv.py (hunk 5 above)
Comment (maxspahn) [reply planarenvs/planarCommon/planarEnv.py:89]: Any news on that?

**Review by maxspahn — CHANGES_REQUESTED** [review]

File: planarenvs/planarCommon/planarEnv.py (hunk 5 above)
Comment (alxschwrz) [reply planarenvs/planarCommon/planarEnv.py:89]: @maxspahn where do you see the potential conflict? I can split it into another state of course (e.g. sensorState). However is the Sensor information not also part of the state space?

File: sensors/SensorCommon.py (hunk 16 above)
Comment (alxschwrz) [reply sensors/SensorCommon.py]: done

File: planarenvs/planarCommon/planarEnv.py (hunk 5 above)
Comment (maxspahn) [reply planarenvs/planarCommon/planarEnv.py:89]: The state is passed to the integration scheme. At some future point, we might want to check the size of the state that is passed into the continuous dynamic (or the integrated one) to raise errors if something is going wrong there. Such an assertion would fail if the size of the state is different depending on how many sensors there are. 
For me, the state is the state of the robot, although I know that this is not really true in the RL-community.

File: sensors/GoalSensor.py (hunk 15 above)
Comment (maxspahn) [reply sensors/GoalSensor.py]: Let me know when you have it at a state that all the other issues and remarks are addressed. Then I will propose something to generalize it to generic kinematic chains.

File: sensors/SensorCommon.py (hunk 9 above)
Comment (alxschwrz) [reply sensors/SensorCommon.py:33]: sounds good to me 😄

File: planarenvs/planarCommon/planarEnv.py (hunk 5 above)
Comment (alxschwrz) [reply planarenvs/planarCommon/planarEnv.py:89]: Okay I understand. I thought it might be sufficient, since it is easy to access the relevant keys in the state now using `'x'` or `'xdot'` as it is a dictionary. 
Should I go with the option to include an extra "sensorState" variable? I would then add the "state" with the "sensorState" in the function [_get_ob()](https://github.com/alxschwrz/gym_envs_planar/blob/5727bf767cffb6bec36f807f48f68917383f5081/planarenvs/planarCommon/planarEnv.py#L101) in order to return the observation.

**Review by maxspahn — APPROVED** [review]
Everything looks great. Let's merge that, and add the forward kinematics in a different branch. 
Then you can start working with this sensor.
